src/main/scala/chisel3/util/experimental/ForceNames.scala

```diff
@@ -0,0 +1,290 @@
+      def toFirrtl = {
+        val t = instance.toAbsoluteTarget
+        ForceNameAnnotation(t, name)
+      }
+      override def transformClass: Class[_ <: Transform] = classOf[ForceNamesTransform]
+    })
+  }
+
+  /** Force the name of this instance to the name its given during Chisel compilation
+    * This will rename after potential renames from other Custom transforms during FIRRTL compilation
+    * @param instance Signal to name
+    */
+  def apply(instance: chisel3.experimental.BaseModule): Unit = {
+    annotate(new ChiselAnnotation with RunFirrtlTransform  {
+      def toFirrtl = {
+        val t = instance.toAbsoluteTarget
+        ForceNameAnnotation(t, instance.instanceName)
+      }
+      override def transformClass: Class[_ <: Transform] = classOf[ForceNamesTransform]
+    })
+  }
+}
+
+/** Links the user-specified name to force to, with the signal/instance in the FIRRTL design
+  * @throws CustomTransformException when the signal is renamed to >1 target
+  * @param target signal/instance to force the name
+  * @param name name to force it to be
+  */
+case class ForceNameAnnotation(target: IsMember, name: String)
+    extends SingleTargetAnnotation[IsMember] {
+  def duplicate(n: IsMember): ForceNameAnnotation = this.copy(target = n, name)
+
+  // Errors if renaming to multiple targets
+  override def update(renames: RenameMap): Seq[Annotation] = {
+    renames.get(target) match {
+      case None => List(this)
+      case Some(newTargets) if newTargets.size > 1 =>
+        throw CustomTransformException(
+          new FirrtlUserException(
+            s"Cannot force the name of $target to $name because it is renamed to $newTargets." +
+              " Perhaps $target is not a ground type?"
+          )
+        )
+      case Some(newTargets) => newTargets.map(t => duplicate(t))
+    }
+  }
+}
+
+/** Contains utility functions for [[ForceNamesTransform]] */
+object ForceNamesTransform {
+  /** Returns the [[IsModule]] which is referred to, or if a [[ReferenceTarget]], the enclosing [[IsModule]]
+    * @param a signal/instance/module
+    * @return referring IsModule
+    */
+  def referringIsModule(a: IsMember): IsModule = a match {
+    case b: ModuleTarget    => b
+    case b: InstanceTarget  => b.targetParent
+    case b: ReferenceTarget => b.copy(component = Nil).targetParent.asInstanceOf[IsModule]
+  }
+
+  /** Returns a function which returns all instance paths to a given IsModule
+    * @param graph
+    * @return
+    */
+  def allInstancePaths(graph: InstanceKeyGraph): IsModule => List[List[(Instance, OfModule)]] = {
+    val lookup: String => List[List[(Instance, OfModule)]] =
+      str => graph.findInstancesInHierarchy(str)
+        .view
+        .map(_.map(_.toTokens).toList)
+        .toList
+    allInstancePaths(lookup) _
+  }
+
+  /** Returns a function which returns all instance paths to a given IsModule
+    * Uses Lists because we want to recurse on inner List (head :: tail)-style
```

Comment:
ScalaDoc style, the first line should be a short summary, followed by a blank comment line, followed by longer description. This is just a very long description.

Reply:
Also this justification for the use of `List` doesn't really bear out in the uses below. `List` is good but I would drop the justification.